out_writeapi_test.go

```diff
@@ -491,3 +489,350 @@ func TestFLBPluginFlushCtx(t *testing.T) {
+
+			assert.Equal(t, "FOO", textField.String())
+			assert.Equal(t, "000", timeField.String())
+
+			return nil, nil
+		},
+		FinalizeFunc: func(ctx context.Context, opts ...gax.CallOption) (int64, error) {
+			return 0, nil
+		},
+		FlushRowsFunc: func(ctx context.Context, offset int64, opts ...gax.CallOption) (int64, error) {
+			return 0, nil
+		},
+		StreamNameFunc: func() string {
+			return ""
+		},
+	}
+
+	origFunc := getWriter
+	getWriter = func(client ManagedWriterClient, ctx context.Context, projectID string, opts ...managedwriter.WriterOption) (MWManagedStream, error) {
+		return mockMS, nil
+	}
+	defer func() { getWriter = origFunc }()
+
+	patch1 := monkey.Patch(output.FLBPluginConfigKey, func(plugin unsafe.Pointer, key string) string {
+		switch key {
+		case "Exactly_Once":
+			return "True"
+		default:
+			return ""
+		}
+	})
+	defer patch1.Unpatch()
+
+	patchSetContext := monkey.Patch(output.FLBPluginSetContext, func(plugin unsafe.Pointer, ctx interface{}) {
+		setID = ctx.(int)
+	})
+	defer patchSetContext.Unpatch()
+
+	initRes := FLBPluginInit(nil)
+
+	orgFunc := getFLBPluginContext
+	getFLBPluginContext = func(ctx unsafe.Pointer) int {
+		checks.calledGetContext++
+		return setID
+	}
+	defer func() { getFLBPluginContext = orgFunc }()
+
+	origReadyFunc := isReady
+	isReady = func(queueHead *managedwriter.AppendResult) bool {
+		// Response is always ready for test
+		return true
+	}
+	defer func() { isReady = origReadyFunc }()
+
+	origResultFunc := pluginGetResult
+	pluginGetResult = func(queueHead *managedwriter.AppendResult, ctx context.Context) (int64, error) {
+		checks.getResultsCount++
+		return -1, nil
+	}
+	defer func() { pluginGetResult = origResultFunc }()
+
+	patchDecoder := monkey.Patch(output.NewDecoder, func(data unsafe.Pointer, length int) *output.FLBDecoder {
+		checks.createDecoder++
+		return nil
+	})
+	defer patchDecoder.Unpatch()
+
+	var rowSent int = 0
+	var rowCount int = 5
+	patchRecord := monkey.Patch(output.GetRecord, func(dec *output.FLBDecoder) (ret int, ts interface{}, rec map[interface{}]interface{}) {
+		checks.gotRecord++
+		dummyRecord := make(map[interface{}]interface{})
+		if rowSent < rowCount {
+			rowSent++
+			// Represents "FOO" in bytes as the data for the Text field
+			dummyRecord["Text"] = []byte{70, 79, 79}
+			// Represents "000" in bytes as the data for the Time field
+			dummyRecord["Time"] = []byte{48, 48, 48}
+			return 0, nil, dummyRecord
+		}
+		// Reset to prepare for next call to Flush
+		rowSent = 0
+		return 1, nil, nil
+	})
+	defer patchRecord.Unpatch()
+
+	// Converts id (int) to type unsafe.Pointer to be used as the ctx
+	uintptrValue := uintptr(setID)
+	pointerValue := unsafe.Pointer(uintptrValue)
+
+	// Calls FlushCtx with this ID
+	result := FLBPluginFlushCtx(pointerValue, nil, 0, nil)
+	//expect the offset to equal the number of successful rows previously sent (with no server-side errors)
+	assert.Equal(t, int64(rowCount), configMap[setID].offsetCounter)
```

Comment:
Rather than verify using an internal structure of the plugin (which is liable to change), you should be able to verify this by checking the offset value supplied in a subsequent call to AppendRows.

Reply:
Done with test-only method.